srml/rolling-window/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Rolling Window module
+//!
+//! ## Overview
+//!
+//! The Rolling Window Module is similar to `simple moving average` except
+//! that it just reports the number of occurrences in the window instead of
+//! calculating the average.
+//!
+//! It is mainly implemented to keep track of misbehaviors and only the take
+//! the last `sessions` of misbehaviors into account.
+//!
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(test)]
+mod mock;
+
+use srml_support::{
+	StorageValue, StorageMap, EnumerableStorageMap, decl_module, decl_storage,
+	traits::WindowLength
+};
+use parity_codec::Codec;
+use sr_primitives::traits::MaybeSerializeDebug;
+
+type Session = u32;
+
+/// Rolling Window trait
+pub trait Trait: system::Trait {
+	/// Kind to report with window length
+	type Kind: Copy + Clone + Codec + MaybeSerializeDebug + WindowLength<u32>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as RollingWindow {
+		/// Misbehavior reports
+		///
+		/// It maps each kind into a hash (identity of reporter) and session number when it occurred
+		MisconductReports get(kind): linked_map T::Kind => Vec<(Session, T::Hash)>;
+
+		/// Session index
+		SessionIndex get(s) config(): u32 = 0;
+	}
+}
+
+decl_module! {
+	/// Rolling Window module
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {}
+}
+
+impl<T: Trait> Module<T> {
+	/// Remove items that doesn't fit into the rolling window
+	/// Must be called when a session ends
+	pub fn on_session_end() {
+		let session = SessionIndex::get().wrapping_add(1);
+
+		for (kind, _) in <MisconductReports<T>>::enumerate() {
+			let window_length = kind.window_length();
+			<MisconductReports<T>>::mutate(kind, |reports| {
+				// it is guaranteed that `reported_session` happened before `session`
+				reports.retain(|(reported_session, _)| {
```

Comment:
this will continue iterating after passing all items to be pruned away. probably can be done more efficiently with a `position` and `drain` combo. but we can file that for later

Reply:
I'm not sure how to do that because we not really interested in `session_index` instead `session_expire (session_index + window_length)` is relevant. Subsequent items can have very different `session_expire`.

However, alternatively we would store `session_expire` instead but then we couldn't detect whether it `wrapped around or not` because we wouldn't know if `current_session` happened before or after `session_expire`. 

I will file an issue let's revisit this then (it is an TODO atm)